parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -1069,27 +1073,80 @@ pushGitRootBranch srcConn branch repo = runExceptT @GitError do
+      else Nothing
+      where
+        statusLines = Text.unpack <$> Text.lines status
+        okLine ('?':'?':(trimStart -> ".unison/v2/unison.sqlite3")) = True
+        okLine ('M':(trimStart -> ".unison/v2/unison.sqlite3")) = True
+        okLine line = isWalDelete line || isShmDelete line
+        isWalDelete ('D':(trimStart -> ".unison/v2/unison.sqlite3-wal")) = True
+        isWalDelete _ = False
+        isShmDelete ('D':(trimStart -> ".unison/v2/unison.sqlite3-shm")) = True
+        isShmDelete _ = False
+        hasDeleteWal = any isWalDelete statusLines
+        hasDeleteShm = any isShmDelete statusLines
+        trimStart = dropWhile Char.isSpace
+
     -- Commit our changes
     push :: CodebasePath -> RemoteRepo -> IO Bool -- withIOError needs IO
     push remotePath (GitRepo url gitbranch) = time "SqliteCodebase.pushGitRootBranch.push" $ do
       -- has anything changed?
-      status <- gitTextIn remotePath ["status", "--short"]
+      -- note: -uall recursively shows status for all files in untracked directories
+      --   we want this so that we see
+      --     `??  .unison/v2/unison.sqlite3` and not
+      --     `??  .unison/`
+      status <- gitTextIn remotePath ["status", "--short", "-uall"]
       if Text.null status
         then pure False
-        else do
-          gitIn remotePath ["add", "--all", "."]
-          gitIn
-            remotePath
-            ["commit", "-q", "-m", "Sync branch " <> Text.pack (show $ Branch.headHash branch)]
-          -- Push our changes to the repo
-          case gitbranch of
-            Nothing -> gitIn remotePath ["push", "--quiet", url]
-            Just gitbranch ->
-              error $
-                "Pushing to a specific branch isn't fully implemented or tested yet.\n"
-                  ++ "InputPatterns.parseUri was expected to have prevented you "
-                  ++ "from supplying the git treeish `"
-                  ++ Text.unpack gitbranch
-                  ++ "`!"
-              -- gitIn remotePath ["push", "--quiet", url, gitbranch]
-          pure True
+        else case parseStatus status of
+          Nothing ->
+            error $ "An error occurred during push.\n"
+                 <> "I was expecting only to see .unison/v2/unison.sqlite3 modified, but saw:\n\n"
+                 <> Text.unpack status <> "\n\n"
+                 <> "Please visit https://github.com/unisonweb/unison/issues/2063\n"
+                 <> "and add any more details about how you encountered this!\n"
+          Just (hasDeleteWal, hasDeleteShm) -> do
+            -- Only stage files we're expecting; don't `git add --all .`
+            -- which could accidentally commit some garbage
+            gitIn remotePath ["add", ".unison/v2/unison.sqlite3"]
+            when hasDeleteWal $ gitIn remotePath ["rm", ".unison/v2/unison.sqlite3-wal"]
+            when hasDeleteShm $ gitIn remotePath ["rm", ".unison/v2/unison.sqlite3-shm"]
```

Comment:
This is the only relevant diff. The rest is just needing to introduce an extra level of indentation for the pattern match. The logic in the `Just` case is otherwise the same as before.